File: python/perfect-numbers/perfect_numbers.py
import collections
import itertools

def classify(number):
    if number < 1:
        raise ValueError("Not a positive integer")
    # first solution way too slow: ~7secs
    # sum_of_divisors = sum(filter(lambda divisor: number % divisor == 0, range(1,int(number/2 + 1))))

    # solution using prime factors: ~.05secs
    # harder to read and longer, but 2 orders of magnitude faster
    divisors = list(get_divisors(number))
    sum_of_divisors = sum(divisors[0:len(divisors)-1])
    if sum_of_divisors == number:
        return "perfect"
    elif sum_of_divisors > number:
        return "abundant"
    else:
        return "deficient"

def prime_factors(number):
    possible_prime_factor = 2
    while possible_prime_factor * possible_prime_factor <= number:
        if number % possible_prime_factor == 0:
            number /= possible_prime_factor 
            yield possible_prime_factor
        else:
            possible_prime_factor += 1

    if number>1:
        yield number


def product(iterable):
    prod = 1
    for factor in iterable:
        prod *= factor
    return prod


def get_divisors(number):
    pf = prime_factors(number)
    pf_with_counts = collections.Counter(pf)

    powers_of_pf = [
            [factor ** exponent for exponent in range(count+1)]
            for factor, count in pf_with_counts.items()
    ]
    for pf_powers_combo in itertools.product(*powers_of_pf):
        yield product(pf_powers_combo)

File: javascript/perfect-numbers/perfect-numbers.js
export const classify = (n) => {
  if (n < 1) { throw new Error('Classification is only possible for natural numbers.')}
  const factors = get_factors(n)
  const sum_of_factors = factors.reduce((sum, f) => sum+=f, 0)
  if (sum_of_factors == n) {
    return "perfect"
  } else if (sum_of_factors > n) {
    return "abundant"
  } else {
    return "deficient"
  }
};

function prime_factors(n) {
  let possible_prime = 2
  let prime_factors = []
  while (possible_prime**2 <= n) {
    if (n%possible_prime == 0) {
      n/=possible_prime
      prime_factors.push(possible_prime)
    } else {
      possible_prime+=1
    }
  }
  if (n > 1) { prime_factors.push(n) }
  return prime_factors
}

function prime_factors_count(prime_factors) {
  return prime_factors.reduce((counts, pf) => {
    counts[pf] = (counts[pf] || 0) + 1;
    return counts;
  }, {});
}

function range(n) {
  return Array(n).fill(1).map((_, i) => i);
}

function get_factors(n) {
  const pfs = prime_factors(n);
  const pf_counts = prime_factors_count(pfs);

  let factors = [];
  for (let [factor, count] of Object.entries(pf_counts)) {
    for (let exponent of range(count+1)) {
      if (!factors.includes(factor**exponent)) {
        factors.push(factor ** exponent)
      }
    }
  }

  factors.forEach(f1 => {
    factors.forEach(f2 => {
      if (f1*f2 < n && n%(f1*f2)==0 && !factors.includes(f1*f2)) {
        factors.push(f1*f2)
      }
    })
  })
  return factors.filter(f => f!=n)
}

